internal/app/singularity/pull.go

```diff
@@ -68,11 +68,19 @@ func LibraryPull(imgCache *cache.Handle, name, ref, transport, fullURI, libraryU
 
 	imagePath := imgCache.LibraryImage(libraryImage.Hash, imageName)
 	exists, err := imgCache.LibraryImageExists(libraryImage.Hash, imageName)
-	if err != nil {
+	if err == cache.ErrBadChecksum {
+		sylog.Warningf("Removing cached image: %s: cache could be corrupted", imagePath)
+		err := os.Remove(imagePath)
+		if err != nil {
+			return fmt.Errorf("unable to remove corrupted cache: %v", err)
+		}
+	} else if err != nil {
 		return fmt.Errorf("unable to check if %s exists: %v", imagePath, err)
 	}
+
 	if !exists {
 		sylog.Infof("Downloading library image")
+		go InterruptCleanup(imagePath)
 
 		// call library download image helper
 		if err = library.DownloadImage(context.TODO(), libraryClient, imagePath, imageRef, downloadImageCallback); err != nil {
```

Comment:
since this is using a context, the signal handler should use it to notify the download process that it should stop. I would expect that if the download is stopped, this functions returns an error, and you can let the normal code flow continue from here (dealing with the error).

This also hints at the signal handler being installed "too low". It should probably be installed higher in the call stack so that you can pass a context down and handle cancellation in that way, as it is likely that all clients need the same handling.

Reply:
This is a good idea, but I think it should be added as an additional improvement (see #3899).  Since this is fixing a bug, I think it can be added as it already is, and the context can be fixed later.